### ISSUE TITLE
Provide a fluent syntax to configure successive calls to the same member.

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#803]
Currently, configuring successive calls to the same member is a bit counter-intuitive, because the configuration order is the reverse of the call order. For instance, if I do this:

```
A.CallTo(() => foo.Bar()).Returns(1);
A.CallTo(() => foo.Bar()).Throws<InvalidOperationException>().Once();
A.CallTo(() => foo.Bar()).Returns(0).Twice();
```

It will return 0 on the first two calls, then throw an exception, then return 1 forever. Even though it makes sense from an implementation perspective, this behavior is very surprising and confusing to people using FakeItEasy for the first time, and there's no way you can guess it works like this without looking at the [documentation](http://fakeiteasy.readthedocs.io/en/stable/limited-call-specifications/#changing-behavior-between-calls).

I'm not suggesting we change the current behavior, because too much existing code relies on it, but we could provide a more intuitive syntax:

```
A.CallTo(() => foo.Bar()).Returns(0).Twice()
    .Then.Throws<InvalidOperationException>().Once()
    .Then.Returns(1);
```

I know that having two ways to do the same thing is something we usually try to avoid, but in this case I think it's worth it, because:
- it makes the code less surprising and easier to understand, especially for newcomers
- it's more discoverable
- as a side benefit, the call specification no longer needs to be repeated.

The return type of `Once`/`Twice`/`NumberOfTimes` would have to change from `void` to something else; this would be a breaking change at the binary level, but not at the source level, except in the (unlikely) case where one of these methods is assigned to a delegate.

`Then` would only be available after a repeat configuration, since it doesn't make sense for a rule that is applied forever.

**Comment by blairconrad** [comment]
Aside from my general reticence to have more than one way to do something, I like the new syntax. Nice idea, @thomaslevesque.
(And the syntax could achieve "blessed" status in the docs, essentially becoming "the way").

I pause a little because we'll still have to maintain the old stack-based behaviour for the "ambient" or "implicit" configuration. That is, when a call is configured using an `IFakeOptionsBuilder` or explicit fake creation options, those configurations will still hang around, to be overridden by what is essentially a queue-based syntax.

I worry that users (many of whom probably don't use the options builders or the explicit fake creation options, so the point could be moot) might be a little confused by the two paradigms, but that's probably a minor concern.
If nobody else thinks this will introduce additional confusion, I'm in.

**Comment by thomaslevesque** [comment]
I don't think the interaction with fake option builders is going to be an issue.

If you do this:

```
A.CallTo(() => foo.Bar()).Returns(0).Twice()
    .Then.Throws<InvalidOperationException>().Once()
    .Then.Returns(1);
```

the `Returns(1)` is "forever", so it's never going to fall back to the default behavior (whether it's the "real" default, or the behavior defined by the options builder)

And if you do this:

```
A.CallTo(() => foo.Bar()).Returns(0).Twice()
    .Then.Throws<InvalidOperationException>().Once()
    .Then.Returns(1).Once();
```

after the configured calls have been "consumed", the fake will fall back to the default behavior. Whether it's the "real" default, or the behavior defined by the options builder, doesn't really matter, and the user probably won't even think about it.

Now that I think of it, the stack-based behavior does seem natural when considering `IFakeOptionsBuilder`, because it's easy to understand that new call configurations will override the ones defined by the options builder. But when you want to configure successive calls within a test, you don't think in terms of overriding (well, _I_ don't, anyway), so using overrides to configure successive calls forces the user to think backward.

**Comment by blairconrad** [comment]
> Now that I think of it, the stack-based behavior does seem natural when considering `IFakeOptionsBuilder`, because it's easy to understand that new call configurations will override the ones defined by the options builder.

Yup. That's how I always explain it. The last thing in wins because it's more likely that you've seen it.

> But when you want to configure successive calls within a test, you don't think in terms of overriding (well, I don't, anyway), so using overrides to configure successive calls forces the user to think backward.

Also true.

**Comment by adamralph** [comment]
I like it. Specially if we promote it in the docs as the *primary* way to configure multiple things on the same fake. The stack based behaviour of multiple configurations can be reduced to a footnote, with a recommendation that you always use the new syntax instead.

**Comment by blairconrad** [comment]
Uncharacteristically, I'd consider removing the **breaking** from this. I know it's technically a signature change, but I'd be stunned if there's any harm in it.

**Comment by thomaslevesque** [comment]
I'm looking into it now, and I'm afraid it's going to be a *lot* harder that I anticipated... The main issue is that `Then` should return the right configuration interface (`IReturnTypeConfiguration<TMember>`, `IVoidConfiguration`, etc), but this information is already lost in `IAfterCallSpecifiedConfiguration`. To propagate it, `IAfterCallSpecifiedConfiguration` would need to become generic; but that, in turn, would require `IExceptionThrowerConfiguration`, `ICallBaseConfiguration`, `IOutAndRefParametersConfiguration`, etc, to become generic as well.

It might be feasible, but at the cost of cluttering the API significantly more that I thought. This would make the "minor breaking change" not so minor...

**Comment by adamralph** [comment]
If, for a moment, we pretended that FIE had never been released and were working towards 1.0, would we want to build this? I.e. is it the compelling API that we'd want to have straight away if we'd thought of it?

What I'm getting at, is that it if this is were to be released in a new major, I'm not sure it really matters "how" breaking it is if it's the way we think things ought to be. Given the amount of change involved, I'd want to be sure of that. I.e. we'd have to be sure that the new syntax would cover close to all use cases and would be more desirable than the alternative in close to all of them.

**Comment by thomaslevesque** [comment]
Well, I can try to POC it, and if we don't like the result, we don't have to use it...

**Comment by adamralph** [comment]
Are we still POC'ing this? I.e. not sure if we actually want to do it?

**Comment by thomaslevesque** [comment]
> Are we still POC'ing this?

As far as I'm concerned, the POC is done, and working (see #804). Waiting for your comments :wink:

> I.e. not sure if we actually want to do it?

Well, I'm sure *I* want to do it... now I just have to convince you :wink:

**Comment by blairconrad** [comment]
POC seemed good. I am bullish on the API. I've just been neglecting the PR because of .NET Standard and other stuff. Sorry, @thomaslevesque.

**Comment by thomaslevesque** [comment]
> I've just been neglecting the PR because of .NET Standard and other stuff. Sorry, @thomaslevesque.

No problem, I realize it's not the highest priority at the moment

**Comment by adamralph** [comment]
In that case, I'm happy to proceed with this and release it with 3.0. Let's not merge the PR into master just yet though, until we agree how we are going to handle the .NET core release (3.0 or not) and how we want to handle the branching if it is part of 3.0.

**Comment by adamralph** [comment]
Removed from the 3.0 milestone. We'll assign it to the next relevant milestone if/when the PR gets merged.

**Comment by blairconrad** [comment]
This issue has been released in FakeItEasy 3.0.0:
https://github.com/FakeItEasy/FakeItEasy/releases/3.0.0
